packages/devtools_app/lib/src/shared/routing.dart

```diff
@@ -328,6 +340,16 @@ class DevToolsRouterDelegate extends RouterDelegate<DevToolsRouteConfiguration>
     }
     return currentState.hasChanges(changes);
   }
+
+  /// Connects to the VM Service if it is not already connected.
+  Future<void> _maybeConnectToVmService(String vmServiceUri) async {
+    final alreadyConnected =
+        serviceConnection.serviceManager.connectedState.value.connected;
+    // Skip connecting if we are already connected or in a test environment.
+    if (!alreadyConnected && !_isTestMode) {
+      await FrameworkCore.initVmService(serviceUriAsString: vmServiceUri);
+    }
```

Comment:
nit: this may be easier to follow with an early return
`if (alreadyConnected || isTestMode) return;`